### ISSUE TITLE
Make `LogitsProcessor`s compute the next mask asynchronously

### ISSUE THREAD

**Issue by rlouf** [dottxt-ai/outlines#1434]
Building the mask is currently blocking inference. However this does not need to be: the only operation that is necessarily blocking is applying the mask to the logits. Everything else can be done during the forward pass. I thus suggest we modify the implementation of logits processors so the next token mask is computed and moved to device asynchronously. Currently blocked by https://github.com/dottxt-ai/outlines-core/issues/178.

**Comment by rlouf** [comment]
We could also attempt to perform compilation during the first forward pass that typically takes longer because of prompt processing. However, since computing index can be slow we still want to leave users the option to pass a `Guide` directly.
